### PR TITLE
Major Refactoring

### PR DESCRIPTION
**Pull request by lucasborin**
- solves #295
- probably solves #296
- solves #298
- solves #299
- a few pseudo comments were moved 
- new attributes to handle relevant statement/structure types
- execute_check split in inspect_structures and inspect_statements

### CODE REVIEW

File: src/checks/y_check_form.clas.testclasses.abap [hunk 1]
```diff
@@ -50,9 +50,9 @@ CLASS local_test_class IMPLEMENTATION.
       ( ' START-OF-SELECTION. ' )
       ( '   PERFORM example. ' )
 
-      ( ' FORM example. ' )
+      ( ' FORM example. "#EC CI_FORM ' )
       ( '   DATA class TYPE REF TO cl_oo_class.' )
-      ( ' ENDFORM. "#EC CI_FORM' )
+      ( ' ENDFORM. ' )
```
Comment (lucasborin) [inline src/checks/y_check_form.clas.testclasses.abap:55]: Document this change.

File: src/checks/y_check_is_interface_in_class.clas.testclasses.abap [hunk 2]
```diff
@@ -76,8 +76,8 @@ CLASS ltc_class IMPLEMENTATION.
     result = VALUE #(
       ( 'REPORT ut_repo.' )
 
-      ( 'CLASS lcl_classname DEFINITION. "#EC INTF_IN_CLASS' )
-      ( ' PUBLIC SECTION.' )
+      ( 'CLASS lcl_classname DEFINITION. ' )
+      ( ' PUBLIC SECTION. "#EC INTF_IN_CLASS' )
```
Comment (lucasborin) [inline src/checks/y_check_is_interface_in_class.clas.testclasses.abap:80]: Document this change.

File: src/checks/y_check_num_public_attributes.clas.testclasses.abap [hunk 3]
```diff
@@ -52,8 +52,8 @@ CLASS ltc_instance IMPLEMENTATION.
     result = VALUE #(
       ( 'REPORT y_example. ' )
 
-      ( ' CLASS y_example DEFINITION. "#EC NUM_PUBLIC_ATTR ' )
-      ( '   PUBLIC SECTION. ' )
+      ( ' CLASS y_example DEFINITION.  ' )
+      ( '   PUBLIC SECTION. "#EC NUM_PUBLIC_ATTR ' )
```
Comment (lucasborin) [inline src/checks/y_check_num_public_attributes.clas.testclasses.abap:56]: Document it as well

**Comment by larshp** [comment]
looks like the 5 issues found by abaplint are correct? they are syntax errors,

image

**Comment by lucasborin** [comment]
Yes, there are a few objects to be fixed as it is a work in progress (draft pull-request). 
I have created it to review the current status.

File: src/foundation/y_check_base.clas.abap [hunk 4]
```diff
@@ -225,32 +223,31 @@ CLASS y_check_base IMPLEMENTATION.
...
       DATA(index) = <structure>-stmnt_from.
 
-      LOOP AT ref_scan_manager->get_statements( ) ASSIGNING FIELD-SYMBOL(<statement>)
-        FROM <structure>-stmnt_from TO <structure>-stmnt_to.
+      LOOP AT ref_scan_manager->statements ASSIGNING FIELD-SYMBOL(<statement>)
+      FROM <structure>-stmnt_from
+      TO <structure>-stmnt_to.
+        IF is_in_scope( <statement> ) = abap_false.
+          CONTINUE.
+        ENDIF.
```
Comment (lucasborin) [inline src/foundation/y_check_base.clas.abap:243]: Will it impact the index?

File: src/checks/y_check_boolean_input_param.clas.abap [hunk 5]
```diff
@@ -22,35 +25,12 @@ CLASS y_check_boolean_input_param IMPLEMENTATION.
     settings-threshold = 0.
     settings-documentation = |{ c_docs_path-checks }boolean-input-parameter.md|.
 
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-class_definition ) ).
+    relevant_structure_types = VALUE #( ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_boolean_input_param.clas.abap:29]: initial?

File: src/checks/y_check_check_in_loop.clas.abap [hunk 6]
```diff
@@ -25,35 +24,10 @@ CLASS y_check_check_in_loop IMPLEMENTATION.
...
-  ENDMETHOD.
-
-
-  METHOD execute_check.
-    LOOP AT ref_scan_manager->get_structures( ) ASSIGNING FIELD-SYMBOL(<structure>)
-    WHERE stmnt_type EQ scan_struc_stmnt_type-check.
-
-      is_testcode = test_code_detector->is_testcode( <structure> ).
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-check ) ).
+    relevant_structure_types = VALUE #( ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_check_in_loop.clas.abap:28]: initial?

File: src/checks/y_check_comment_usage.clas.testclasses.abap [hunk 7]
```diff
@@ -44,3 +44,66 @@ CLASS local_test_class IMPLEMENTATION.
...
+      ( '*"*COMMENT ' )
+      ( '*" ' )
+      ( '      "! docu ' )
+      ( '    ENDMETHOD.' )
+      ( '  ENDCLASS. ' )
+    ).
+  ENDMETHOD.
+
+  METHOD get_code_with_exemption.
+    result = VALUE #( ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_comment_usage.clas.testclasses.abap:106]: ?

File: src/checks/y_check_constants_interface.clas.abap [hunk 8]
```diff
@@ -43,57 +29,63 @@ CLASS Y_CHECK_CONSTANTS_INTERFACE IMPLEMENTATION.
     settings-threshold = 0.
     settings-documentation = |{ c_docs_path-checks }constants-interface.md|.
 
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-interface ) ).
+    relevant_structure_types = VALUE #( ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_constants_interface.clas.abap:33]: ?

File: src/checks/y_check_db_access_in_ut.clas.abap [hunk 9]
```diff
@@ -56,42 +55,18 @@ CLASS y_check_db_access_in_ut IMPLEMENTATION.
...
-
-      IF is_testcode = abap_false.
-        CONTINUE.
-      ENDIF.
-
-      IF has_osql_or_cds_framework( <structure> ) = abap_true.
-        CONTINUE.
-      ENDIF.
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-method ) ).
+    relevant_structure_types = VALUE #( ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_db_access_in_ut.clas.abap:59]: ?

File: src/checks/y_check_empty_if_branches.clas.abap [hunk 10]
```diff
@@ -85,41 +84,17 @@ CLASS Y_CHECK_EMPTY_IF_BRANCHES IMPLEMENTATION.
...
-  ENDMETHOD.
-
-
-  METHOD execute_check.
-    LOOP AT ref_scan_manager->get_structures( ) ASSIGNING FIELD-SYMBOL(<structure>)
-      WHERE stmnt_type EQ scan_struc_stmnt_type-if.
-
-      is_testcode = test_code_detector->is_testcode( <structure> ).
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-if ) ).
+    relevant_structure_types = VALUE #( (  ) ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_empty_if_branches.clas.abap:88]: ??

File: src/checks/y_check_external_call_in_ut.clas.abap [hunk 11]
```diff
@@ -49,30 +45,11 @@ CLASS Y_CHECK_EXTERNAL_CALL_IN_UT IMPLEMENTATION.
...
-    set_check_message( 'External Call from an Unit-Test should be removed!' ).
-
-  ENDMETHOD.                    "CONSTRUCTOR
-
 
-  METHOD execute_check.
-    LOOP AT ref_scan_manager->get_structures( ) ASSIGNING FIELD-SYMBOL(<structure>)
-      WHERE stmnt_type EQ scan_struc_stmnt_type-method.
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-method ) ).
+    relevant_structure_types = VALUE #( ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_external_call_in_ut.clas.abap:50]: ?

File: src/checks/y_check_form.clas.abap [hunk 12]
```diff
@@ -28,41 +20,27 @@ CLASS Y_CHECK_FORM IMPLEMENTATION.
     settings-threshold = 0.
     settings-documentation = |{ c_docs_path-checks }form-routine.md|.
 
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-form ) ).
+    relevant_structure_types = VALUE #( ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_form.clas.abap:24]: ?

File: src/checks/y_check_is_interface_in_class.clas.abap [hunk 13]
```diff
@@ -31,52 +40,19 @@ CLASS Y_CHECK_IS_INTERFACE_IN_CLASS IMPLEMENTATION.
     settings-apply_on_test_code = abap_false.
     settings-documentation = |{ c_docs_path-checks }interface-in-class.md|.
 
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-public_section ) ).
+    relevant_structure_types = VALUE #(  ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_is_interface_in_class.clas.abap:44]: ?

File: src/checks/y_check_method_output_param.clas.abap [hunk 14]
```diff
@@ -66,46 +60,24 @@ CLASS Y_CHECK_METHOD_OUTPUT_PARAM IMPLEMENTATION.
...
-  ENDMETHOD.
-
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-class_definition )
+                                        ( scan_struc_stmnt_type-interface ) ).
 
-  METHOD execute_check.
-    LOOP AT ref_scan_manager->get_structures( ) ASSIGNING FIELD-SYMBOL(<structure>)
-      WHERE stmnt_type EQ scan_struc_stmnt_type-class_definition OR
-            stmnt_type EQ scan_struc_stmnt_type-interface.
+    relevant_structure_types = VALUE #( ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_method_output_param.clas.abap:66]: ?

File: src/checks/y_check_method_return_bool.clas.abap [hunk 15]
```diff
@@ -29,6 +29,11 @@ CLASS Y_CHECK_METHOD_RETURN_BOOL IMPLEMENTATION.
     settings-prio = c_warning.
     settings-documentation = |{ c_docs_path-checks }method-return-bool.md|.
 
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-class_definition )
+                                        ( scan_struc_stmnt_type-interface ) ).
+
+    relevant_structure_types = VALUE #( ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_method_return_bool.clas.abap:35]: ?

File: src/checks/y_check_num_output_parameter.clas.abap [hunk 16]
```diff
@@ -21,37 +23,16 @@ CLASS Y_CHECK_NUM_OUTPUT_PARAMETER IMPLEMENTATION.
...
-  ENDMETHOD.
-
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-class_definition )
+                                        ( scan_struc_stmnt_type-interface ) ).
 
-  METHOD execute_check.
-    LOOP AT ref_scan_manager->get_structures( ) ASSIGNING FIELD-SYMBOL(<structure>)
-    WHERE stmnt_type EQ scan_struc_stmnt_type-class_definition
-    OR stmnt_type EQ scan_struc_stmnt_type-interface.
+    relevant_structure_types = VALUE #( ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_num_output_parameter.clas.abap:29]: ?

File: src/checks/y_check_num_public_attributes.clas.abap [hunk 17]
```diff
@@ -1,135 +1,103 @@
...
 
-  METHOD checkif_error.
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-public_section ) ).
 
-    DATA(check_configuration) = detect_check_configuration( statement = statement
-                                                            error_count = public_attribute_counter ).
-    IF check_configuration IS INITIAL.
-      RETURN.
-    ENDIF.
+    relevant_structure_types = VALUE #( ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_num_public_attributes.clas.abap:40]: ?

File: src/checks/y_check_number_attributes.clas.abap [hunk 18]
```diff
@@ -1,37 +1,62 @@
...
+    super->constructor( ).
+
+    settings-pseudo_comment = '"#EC NUMBER_ATTR' ##NO_TEXT.
+    settings-threshold = 12.
+    settings-documentation = |{ c_docs_path-checks }number-attributes.md|.
+
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-class_definition )
+                                        ( scan_struc_stmnt_type-interface ) ).
+
+    relevant_structure_types = VALUE #( ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_number_attributes.clas.abap:39]: ?

File: src/checks/y_check_number_events.clas.abap [hunk 19]
```diff
@@ -1,82 +1,44 @@
...
-
-  METHOD execute_check.
-    LOOP AT ref_scan_manager->get_structures( ) ASSIGNING FIELD-SYMBOL(<structure>)
-      WHERE stmnt_type = scan_struc_stmnt_type-class_definition
-         OR stmnt_type = scan_struc_stmnt_type-interface.
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-class_definition )
+                                        ( scan_struc_stmnt_type-interface ) ).
 
-      is_testcode = test_code_detector->is_testcode( <structure> ).
+    relevant_structure_types = VALUE #( ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_number_events.clas.abap:30]: ?

File: src/checks/y_check_number_interfaces.clas.abap [hunk 20]
```diff
@@ -47,43 +25,47 @@ CLASS Y_CHECK_NUMBER_INTERFACES IMPLEMENTATION.
     settings-threshold = 4.
     settings-documentation = |{ c_docs_path-checks }number-interfaces.md|.
 
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-class_definition )
+                                        ( scan_struc_stmnt_type-interface ) ).
+
+    relevant_structure_types = VALUE #( ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_number_interfaces.clas.abap:31]: ?

File: src/checks/y_check_number_methods.clas.abap [hunk 21]
```diff
@@ -1,86 +1,46 @@
...
-  ENDMETHOD.
-
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-class_definition )
+                                        ( scan_struc_stmnt_type-interface ) ).
 
-  METHOD execute_check.
-    LOOP AT ref_scan_manager->get_structures( ) ASSIGNING FIELD-SYMBOL(<structure>)
-      WHERE stmnt_type EQ scan_struc_stmnt_type-class_definition OR
-            stmnt_type EQ scan_struc_stmnt_type-interface.
+    relevant_structure_types = VALUE #( ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_number_methods.clas.abap:32]: ?

File: src/checks/y_check_optional_parameters.clas.abap [hunk 22]
```diff
@@ -20,38 +22,14 @@ CLASS Y_CHECK_OPTIONAL_PARAMETERS IMPLEMENTATION.
     settings-threshold = 0.
     settings-documentation = |{ c_docs_path-checks }optional-parameters.md|.
 
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-class_definition ) ).
+    relevant_structure_types = VALUE #(  ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_optional_parameters.clas.abap:26]: ?

File: src/checks/y_check_prefer_case_to_elseif.clas.abap [hunk 23]
```diff
@@ -37,57 +38,51 @@ CLASS y_check_prefer_case_to_elseif IMPLEMENTATION.
...
-
-
-  METHOD execute_check.
-    DATA counters TYPE counters.
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-if )
+                                        ( scan_struc_stmnt_type-elseif ) ).
 
-    DATA(structures) = ref_scan_manager->get_structures( ).
-    DATA(statements) = ref_scan_manager->get_statements( ).
+    relevant_structure_types = VALUE #( ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_prefer_case_to_elseif.clas.abap:44]: ?

File: src/checks/y_check_prefer_is_not.clas.abap [hunk 24]
```diff
@@ -20,42 +20,15 @@ CLASS y_check_prefer_is_not IMPLEMENTATION.
     settings-threshold = 0.
     settings-documentation = |{ c_docs_path-checks }prefer-is-not-to-not-is.md|.
 
-    set_check_message( 'Prefer IS NOT to NOT IS!' ).
-  ENDMETHOD.
-
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-if ) ).
+    relevant_structure_types = VALUE #(  ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_prefer_is_not.clas.abap:24]: ?

File: src/checks/y_check_pseudo_comment_usage.clas.abap [hunk 25]
```diff
@@ -61,7 +54,16 @@ CLASS Y_CHECK_PSEUDO_COMMENT_USAGE IMPLEMENTATION.
     settings-apply_on_productive_code = abap_true.
     settings-prio = c_note.
 
+    relevant_statement_types = VALUE #( ).
+    relevant_structure_types = VALUE #( ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_pseudo_comment_usage.clas.abap:58]: ??

File: src/checks/y_check_returning_name.clas.abap [hunk 26]
```diff
@@ -21,36 +23,15 @@ CLASS y_check_returning_name IMPLEMENTATION.
...
 
-      TRY.
-          DATA(check_configuration) = check_configurations[ apply_on_testcode = abap_true ].
-        CATCH cx_sy_itab_line_not_found.
-          IF is_testcode EQ abap_true.
-            CONTINUE.
-          ENDIF.
-      ENDTRY.
+    relevant_statement_types = VALUE #( ( scan_struc_stmnt_type-class_definition ) ).
+    relevant_structure_types = VALUE #(  ).
```
Comment (eugen-guenther-sap) [inline src/checks/y_check_returning_name.clas.abap:29]: ?

File: src/checks/y_check_boolean_input_param.clas.abap (hunk 5 above)
Comment (eugen-guenther-sap) [inline src/checks/y_check_boolean_input_param.clas.abap:29]: you overwrite it, with initial?

File: src/checks/y_check_check_in_loop.clas.abap (hunk 6 above)
Comment (eugen-guenther-sap) [inline src/checks/y_check_check_in_loop.clas.abap:28]: initial?

File: src/checks/y_check_boolean_input_param.clas.abap (hunk 5 above)
Comment (eugen-guenther-sap) [inline src/checks/y_check_boolean_input_param.clas.abap:29]: initial?

**Review by eugen-guenther-sap — CHANGES_REQUESTED** [review]
please take a look at the comments

File: src/checks/y_check_comment_usage.clas.testclasses.abap (hunk 7 above)
Comment (lucasborin) [reply src/checks/y_check_comment_usage.clas.testclasses.abap]: Fixed (#4da55f0)

File: src/checks/y_check_boolean_input_param.clas.abap (hunk 5 above)
Comment (lucasborin) [reply src/checks/y_check_boolean_input_param.clas.abap:29]: it is not as the y_check_base has a default value

**Comment by lucasborin** [comment]
y_check_db_access_in_ut results in a timeout => we have to refactor it, or implement the is_in_scope logic in the statement loop as well

**Review by eugen-guenther-sap — APPROVED** [review]

File: src/checks/y_check_form.clas.testclasses.abap (hunk 1 above)
Comment (lucasborin) [reply src/checks/y_check_form.clas.testclasses.abap:55]: #66f4694

File: src/checks/y_check_is_interface_in_class.clas.testclasses.abap (hunk 2 above)
Comment (lucasborin) [reply src/checks/y_check_is_interface_in_class.clas.testclasses.abap:80]: it was documented correctly, but wrong in the source code.

File: src/foundation/y_check_base.clas.abap (hunk 4 above)
Comment (lucasborin) [reply src/foundation/y_check_base.clas.abap]: as far as it was tested locally, no issue so far.

**Pull request merged**